In [98]:
from keras.datasets import mnist
(train_images, train_labels), (test_images,test_labels) = mnist.load_data()

- Mnist 데이터셋은 넘파이 배열의 형태로 케라스에 이미 포함되어 있다
- mnist 문제를 알고리즘이 제대로 작동하는지 확인하기 위한 딥러닝계의 "hello world"이다
- 모델이 학습해야될 훈련세트 구성

In [99]:
train_images.shape

(60000, 28, 28)

In [100]:
train_labels

array([5, 0, 4, ..., 5, 6, 8], dtype=uint8)

In [101]:
test_images.shape

(10000, 28, 28)

In [102]:
test_labels

array([7, 2, 1, ..., 4, 5, 6], dtype=uint8)

1. train_images와 train_labels를 네트워크에 주입
    - 네트워크는 이미지와 레이블을 연관시킬 수 있도록 학습된다
2. test_images에 대한 예측을 네트워크에 요청
    - 이 예측이 맞는지 test_labels확인 한다

In [103]:
from keras import models
from keras import layers

network = models.Sequential()
network.add(layers.Dense(512, activation='relu', input_shape=(28 * 28,)))
network.add(layers.Dense(10,activation='softmax'))

- 신경망의 핵심 구성요소는 일종의 데이터 처리 필터라고 생각할 수 있는 층(layer)이다
- 어떤 데이터가 들어가면 더 유용한 형태로 출력된다
- 구체적으로 층은 주어진 문제에 더 의미있는 표현을 입력 데이터로부터 추출한다
- 이 예에서는 조밀하게 연결된 신경망 층인 Dense 층 2개가 연속되어 있다

---
신경망이 훈련 준비를 마치기 위해서 컴파일 단계에 포함될 세가지가 더 필요하다
1. 손실함수 (loss function) : 훈련 데이터에서 신경망의 성능을 측정하는 방법으로 네트워크가 옳은 방향으로 학습될 수 있도록 도와준다
2. 옵티마이저(optimizer) : 입력된 데이터와 손실 함수를 기반으로 네트워크를 업데이트 하는 메커니즘이다
3. 훈련과 테스트과정을 모니터링할 지표: 여기에서는 정확도만 고려한다

In [104]:
network.compile(optimizer='rmsprop',
                loss='categorical_crossentropy',
                metrics=['accuracy']) # 컴파일 단계

In [105]:
train_images = train_images.reshape((60000,28 * 28))
train_images = train_images.astype('float32') / 255

test_images = test_images.reshape((10000, 28 * 28))
test_images = test_images.astype('float32') / 255 # 이미지 데이터 준비하기

데이터를 네트워크에 맞는 크기로 바꾸고 모든 값을 0과 1 사이로 스케일을 조정  
unit8(0과255의 사이 값) ->float32(0과 1의 사이 값)으로 배열 변경

In [106]:
import keras.utils

In [107]:
# 레이블을 범주형으로 인코딩
from tensorflow.keras.utils import to_categorical

train_labels = to_categorical(train_labels)
test_labels = to_categorical(test_labels)

In [108]:
network.fit(train_images, train_labels, epochs=5, batch_size=128)

Epoch 1/5
469/469 [==============================] - 2s 4ms/step - loss: 0.2497 - accuracy: 0.9288
Epoch 2/5
469/469 [==============================] - 2s 4ms/step - loss: 0.1018 - accuracy: 0.9703
Epoch 3/5
469/469 [==============================] - 2s 4ms/step - loss: 0.0676 - accuracy: 0.9796
Epoch 4/5
469/469 [==============================] - 2s 4ms/step - loss: 0.0488 - accuracy: 0.9854
Epoch 5/5
469/469 [==============================] - 2s 4ms/step - loss: 0.0371 - accuracy: 0.9891


- 케라스에서 fit 메서드를 호출하여 훈련 데이터에 모델을 학습
    - 훈련하는 동안 2개의 정보가 출력된다. 네트워크의 손실과 정확도

In [109]:
test_loss, test_acc = network.evaluate(test_images,test_labels)

313/313 [==============================] - 0s 548us/step - loss: 0.0729 - accuracy: 0.9784


테스트 세트의 정확도는 97.8%로 훈련 세트보다 정확도가 약간은 낮다
- 훈련 정확도와 테스트 정확도 사이의 차이는 과대적합(overfitting)때문
